src/main.ts

```diff
@@ -1,256 +1,152 @@
-};
+export function success<F = never, S = never>(s: S): Dataway<F, S> {
+  return { _tag: 'Success', success: s };
+}
 
-const of = success;
+export function isNotAsked<L, A>(ma: Dataway<L, A>): ma is NotAsked {
+  return ma._tag === 'NotAsked' ? true : false;
+}
+export function isLoading<L, A>(ma: Dataway<L, A>): ma is Loading {
+  return ma._tag === 'Loading' ? true : false;
+}
+export function isFailure<L, A>(ma: Dataway<L, A>): ma is Failure<L> {
+  return ma._tag === 'Failure' ? true : false;
+}
+export function isSuccess<L, A>(ma: Dataway<L, A>): ma is Success<A> {
+  return ma._tag === 'Success' ? true : false;
+}
 
 export const fromEither = <L, A>(e: Either<L, A>) => {
-  if (e.isLeft()) {
-    return failure<L, A>(e.value);
+  if (isLeft(e)) {
+    return failure<L, A>(e.left);
   }
 
-  return success<L, A>(e.value);
+  return success<L, A>(e.right);
 };
 
-const ap = <L, A, B>(
-  fab: Dataway<L, (a: A) => B>,
-  fa: Dataway<L, A>,
-): Dataway<L, B> => fa.ap(fab);
-
-const map = <L, A, B>(fa: Dataway<L, A>, f: (a: A) => B): Dataway<L, B> =>
-  fa.map(f);
-
-const chain = <L, A, B>(
-  fa: Dataway<L, A>,
-  f: (a: A) => Dataway<L, B>,
-): Dataway<L, B> => fa.chain(f);
-
+export const fromOption = <L>(e: L) => <A>(o: Option<A>) => {
+  if (isNone(o)) {
+    return failure(e);
+  }
+  return success(o.value);
+};
 export const map2 = <L, A, B, C>(
   f: (a: A) => (b: B) => C,
   fa: Dataway<L, A>,
   fb: Dataway<L, B>,
-): Dataway<L, C> => ap(fa.map(f), fb);
+): Dataway<L, C> => dataway.ap(dataway.map(fa, f), fb);
 
 export const map3 = <L, A, B, C, D>(
   f: (a: A) => (b: B) => (c: C) => D,
   fa: Dataway<L, A>,
   fb: Dataway<L, B>,
   fc: Dataway<L, C>,
-): Dataway<L, D> => ap(ap(fa.map(f), fb), fc);
+): Dataway<L, D> => dataway.ap(dataway.ap(dataway.map(fa, f), fb), fc);
 
 export const append = <L, A, B>(
   fa: Dataway<L, A>,
   fb: Dataway<L, B>,
 ): Dataway<L, [A, B]> => {
-  return ap(fa.map((a: A) => (b: B): [A, B] => [a, b]), fb);
+  return dataway.ap(dataway.map(fa, (a: A) => (b: B): [A, B] => [a, b]), fb);
 };
 
 export const fold = <L, A, B>(
-  onNotAsked: B,
-  onLoading: B,
+  onNotAsked: () => B,
+  onLoading: () => B,
   onFailure: (e: L) => B,
   onSuccess: (a: A) => B,
-  fa: Dataway<L, A>,
+  ma: Dataway<L, A>,
 ): B => {
-  return fa.fold(onNotAsked, onLoading, onFailure, onSuccess);
+  switch (ma._tag) {
+    case 'NotAsked':
+      return onNotAsked();
+
+    case 'Loading':
+      return onLoading();
+
+    case 'Failure':
+      return onFailure(ma.failure);
+
+    default:
+      return onSuccess(ma.success);
+  }
 };
 
 export const dataway: Monad2<URI> = {
   URI,
-  of,
-  ap,
-  map,
-  chain,
+  of: success,
+  ap: (mab, ma) => {
+    switch (ma._tag) {
+      case 'NotAsked':
+        if (isFailure(mab)) {
+          return mab as any;
+        }
+        return (isNotAsked(mab) ? mab : ma) as any;
+
+      case 'Loading':
+        if (isNotAsked(mab)) {
+          return notAsked;
```

Comment:
```suggestion
          return mab;
```
pour faire comme au dessus ?